ext/mixed/js/japanese.js

```diff
@@ -175,3 +305,145 @@ function jpDistributeFuriganaInflected(expression, reading, source) {
 
     return output;
 }
+
+function jpConvertHalfWidthKanaToFullWidth(text, sourceMapping) {
+    let result = '';
+    const ii = text.length;
+    const hasSourceMapping = Array.isArray(sourceMapping);
+
+    for (let i = 0; i < ii; ++i) {
+        const c = text[i];
+        const mapping = JP_HALFWIDTH_KATAKANA_MAPPING.get(c);
+        if (typeof mapping !== 'string') {
+            result += c;
+            continue;
+        }
+
+        let index = 0;
+        switch (text.charCodeAt(i + 1)) {
+            case 0xff9e: // dakuten
+                index = 1;
+                break;
+            case 0xff9f: // handakuten
+                index = 2;
+                break;
+        }
+
+        let c2 = mapping[index];
+        if (index > 0) {
+            if (c2 === '-') { // invalid
+                index = 0;
+                c2 = mapping[0];
+            } else {
+                ++i;
+            }
+        }
+
+        if (hasSourceMapping && index > 0) {
+            index = result.length;
+            const v = sourceMapping.splice(index + 1, 1)[0];
+            sourceMapping[index] += v;
+        }
+        result += c2;
+    }
+
+    return result;
+}
+
+function jpConvertNumericTofullWidth(text) {
+    let result = '';
+    for (let i = 0, ii = text.length; i < ii; ++i) {
+        let c = text.charCodeAt(i);
+        if (c >= 0x30 && c <= 0x39) { // ['0', '9']
+            c += 0xff10 - 0x30; // 0xff10 = '0' full width
+            result += String.fromCharCode(c);
+        } else {
+            result += text[i];
+        }
+    }
+    return result;
+}
+
+function jpConvertAlphabeticToKana(text, sourceMapping) {
+    let part = '';
+    let result = '';
+    const ii = text.length;
+
+    if (sourceMapping.length === ii) {
+        sourceMapping.length = ii;
+        sourceMapping.fill(1);
+    }
+
+    for (let i = 0; i < ii; ++i) {
+        let c = text.charCodeAt(i);
+        if (c >= 0x41 && c <= 0x5a) { // ['A', 'Z']
+            c -= 0x41;
+        } else if (c >= 0x61 && c <= 0x7a) { // ['a', 'z']
+            c -= 0x61;
+        } else if (c >= 0xff21 && c <= 0xff3a) { // ['A', 'Z'] full width
+            c -= 0xff21;
+        } else if (c >= 0xff41 && c <= 0xff5a) { // ['a', 'z'] full width
+            c -= 0xff41;
+        } else {
```

Comment:
```diff
diff --git a/ext/mixed/js/japanese.js b/ext/mixed/js/japanese.js
index 93366db..d595acc 100644
--- a/ext/mixed/js/japanese.js
+++ b/ext/mixed/js/japanese.js
@@ -384,6 +384,9 @@ function jpConvertAlphabeticToKana(text, sourceMapping) {
             c -= 0xff21;
         } else if (c >= 0xff41 && c <= 0xff5a) { // ['a', 'z'] full width
             c -= 0xff41;
+        } else if (c === 0x2d) { // '-'
+            part += String.fromCharCode(c);
+            continue;
         } else {
             if (part.length > 0) {
                 result += jpToHiragana(part, sourceMapping, result.length);
```

I think something similar to this is worth adding before merge to support '-' as an "alphabetic" substitute for ー. Macron '◌̄' is often used too, but there can be combining and direct variants of those so it would require significant changes, and it's also used to mean あ, い or う in text that's normally written in hiragana. I still approve the PR with or without this change.

Wildcards seem to be separated from deinflections anyway so it's probably too complex to add support for them.

Reply:
I updated the format a little bit in that area and added support for both the standard dash and the fullwidth dash.

Tested with:
```
シーツ
shiitsu
SHIITSU
ｓｈｉｉｔｓｕ
ＳＨＩＩＴＳＵ
shi-tsu
SHI-TSU
ｓｈｉーｔｓｕ
ＳＨＩーＴＳＵ
ｓｈｉ－ｔｓｕ
ＳＨＩ－ＴＳＵ
```

Reply:
Confirmed working as expected. Thanks for the test case when repeated vowels can mean the long vowel mark. Japanese romanization is full of little things like this that make it a pain to work with. People also use a mix of Hepburn, Nihon-shiki/Kunrei and IME shortcuts or conventions like xa/la for ぁ or thi for ティ.